align_system/utils/alignment_utils.py

```diff
@@ -0,0 +1,188 @@
+        3. Call _select_min_dist_choice() and return selected choice and probs
+        '''
+        pass
+
+    def _select_min_dist_choice(self, choices, dists, misaligned=False):
+        if not misaligned:
+            # For aligned, want to minimize to distance to target
+            # Invert distances so minimal distances have higher probability
+
+            # Small epsilon for a perfect (0 distance) match
+            inv_dists = [1/(distance+1e-16) for distance in dists]
+
+            # Convert inverse distances to probabilities
+            probs = [inv_dist/sum(inv_dists) for inv_dist in inv_dists]
+        else:
+            # For misaligned, want to maximize distance to target, so
+            # maximize over non-inverted distances
+
+            # Convert distances to probabilities
+            probs = [dist/sum(dists) for dist in dists]
+
+        max_prob = max(probs)
+        max_actions = [idx for idx, p in enumerate(probs) if p == max_prob]
+        # Randomly chose one of the max probability actions
+        # TODO could add some tie breaking logic here
+        selected_choice = choices[random.choice(max_actions)]
+
+        return selected_choice, probs
+
+
+class AvgDistScalarAlignment(AlignmentFunction):
+    def __call__(self, kdma_values, target_kdmas, misaligned=False): 
+        '''
+        Selects a choice by first averaging score across samples,
+        then selecting the one with minimal MSE to the scalar target.
+        Returns the selected choice.
+        '''
+        kdma_values = _handle_single_value(kdma_values, target_kdmas)
+        _check_if_targets_are_scalar(target_kdmas)
+        
+        # Get distance from average of predicted scores to targets
+        distances = []
+        choices = list(kdma_values.keys())
+        for choice in choices:
+            distance = 0.
+            for target_kdma in target_kdmas:
+                kdma = target_kdma['kdma']
+                samples = kdma_values[choice][kdma]
+                average_score = (sum(samples) / len(samples))
+                distance += _euclidean_distance(target_kdma['value'], average_score)
+            distances.append(distance)
+
+        selected_choice, probs = self._select_min_dist_choice(choices, distances, misaligned)
+
+        return selected_choice, probs
+
+
+class MinDistToRandomSampleKdeAlignment(AlignmentFunction): 
+    def __call__(self, kdma_values, target_kdmas, misaligned=False, kde_norm='globalnorm'):
+        '''
+        Returns the choice with min average distance to random sample from the target KDEs
+        '''
+        _check_if_targets_are_kde(target_kdmas)
+
+        # Sample KDEs to get scalar targets
+        sampled_target_kdmas = []
+        for target_kdma in target_kdmas:
+            sampled_target_kdma = {'kdma':target_kdma.kdma}
+            target_kde = kde_utils.load_kde(target_kdma, kde_norm)
+            sampled_target_kdma['value']= float(target_kde.sample(1)) # sample returns array
```

Comment:
Yeah I removed this option for now for simplicity because as the number of samples increases it approaches the maximum likelihood solution. A single sample works best for multimodal distributions. We can add the option back in down the line if we think there's an instance where it'd be useful